ethclient/ethclient.go

```diff
@@ -76,6 +77,15 @@ type rpcBlock struct {
 	UncleHashes  []common.Hash    `json:"uncles"`
 }
 
+func (ec *Client) BlockNumber(ctx context.Context) (uint64, error) {
+	var number string
```

Comment:
Well, I don't like the numbers when you add another, no, wait, *two* other conversions with `big.Int`. So maybe I should keep the rpcclient then:

```
$ go test -v -cpu 4 . -bench .
goos: linux
goarch: amd64
BenchmarkLoInt-4           	50000000	        20.4 ns/op
BenchmarkHiInt-4           	30000000	        38.2 ns/op
BenchmarkLoBig-4           	10000000	       219 ns/op
BenchmarkHiBig-4           	 5000000	       299 ns/op
BenchmarkLoBigViaInt64-4   	10000000	       188 ns/op
BenchmarkHiBigViaInt64-4   	10000000	       268 ns/op
```

https://gist.github.com/nolash/67e5fc146b911712af44528819dd49c8

Reply:
> I don't like the numbers when you add another, no, wait, two other conversions with big.Int

I'm not sure what you mean, I'm saying pass `big.Int` into the rpc client so that it decodes the response into a `big.Int` (which is that the backend function actually returns) rather than here manually decoding into a `uint64`?

> So maybe I should keep the rpcclient then

Why do you need to keep the rpcclient? This seems like a useful addition to ethclient no?

As for the benchmark, I'm not really sure what you are benchmarking here, are you concerned that decoding into a `big.Int` is slower than into a `uint64`? Why does it matter, surely we are not calling `BlockNumber` that often, it only changes every ~10s?

I'm more concerned with the assumption that `BlockNumber` is a `uint64` when it isn't (it's a `big.Int`).

Reply:
`GetBlock()` is called for every Resource creation, update and lookup for latest update. In the end I'm not how the use cases will be, and how often these calls will be performed.

I see your point that the middleware should reflect the data type of the backend, but I find it a bit bureaucratic to waste 10x extra cycles when it's not necessary: I mean, do you think we will cross the max uint64 blocknumber anytime soon?

Reply:
I see that `rpc.LatestBlockNumber` which the backend function encapulates is type rpc.BlockNumber with is even int64...

I'll look to see if we can make this hook better. It seems to me to be a real waste of resources.

Reply:
To not get too out of context for this PR I'll implement as `big.Int` for nowbut will investigate how to slim it later.

Reply:
> GetBlock() is called for every Resource creation, update and lookup for latest update. 

So then we should cache the return value and invalidate it when a new block is mined (which happens every ~10s), there is no need to continue to call an RPC method that is most likely to return the same value.

> Actually, maybe the backend itself should not be big.Int for the blocknumber...

I guess this is the point I was trying to make, if an RPC method returns `big.Int`, then the client should also return the same thing, it's supposed to look transparent. If we think `big.Int` is wrong, it should be changed consistently everywhere.

Reply:
@nolash doesn't look like this will be accepted anyway: https://github.com/ethereum/go-ethereum/pull/15907